overlord/ifacestate/hotplug.go

```diff
@@ -0,0 +1,248 @@
+	if key == "" || key == ":::" {
+		return false
+	}
+	return true
+}
+
+func hotplugTaskSetAttrs(task *state.Task, deviceKey, ifaceName string) {
+	task.Set("device-key", deviceKey)
+	task.Set("interface", ifaceName)
+}
+
+func hotplugTaskGetAttrs(task *state.Task) (deviceKey, ifaceName string, err error) {
+	if err = task.Get("interface", &ifaceName); err != nil {
+		return "", "", fmt.Errorf("internal error: failed to get interface name: %s", err)
+	}
+	if err = task.Get("device-key", &deviceKey); err != nil {
+		return "", "", fmt.Errorf("internal error: failed to get device key: %s", err)
+	}
+	return deviceKey, ifaceName, err
+}
+
+func defaultDeviceKey(devinfo *hotplug.HotplugDeviceInfo) string {
+	vendor, _ := devinfo.Attribute("ID_VENDOR_ID")
+	model, _ := devinfo.Attribute("ID_MODEL_ID")
+	rev, _ := devinfo.Attribute("ID_REVISION")
+	serial, _ := devinfo.Attribute("ID_SERIAL_SHORT")
+	return fmt.Sprintf("%s:%s:%s:%s", vendor, model, rev, serial)
+}
+
+// HotplugDeviceAdded gets called when a device is added to the system.
+func (m *InterfaceManager) HotplugDeviceAdded(devinfo *hotplug.HotplugDeviceInfo) {
+	st := m.state
+	st.Lock()
+	defer st.Unlock()
+
+	// FIXME: agreement needed how to find about system snap and where to attach interfaces.
+	coreSnapName, err := m.repo.GuessSystemSnapName()
+	if err != nil {
+		logger.Noticef("cannot determine system snap name: %s", err)
+		return
+	}
+	coreSnapInfo, err := snapstate.CurrentInfo(st, coreSnapName)
+	if err != nil {
+		logger.Noticef("%q snap not available, hotplug events ignored", coreSnapName)
+		return
+	}
+
+	hotplugIfaces := m.repo.AllHotplugInterfaces()
+	defaultKey := defaultDeviceKey(devinfo)
+
+	hotplugFeature, err := m.hotplugEnabled()
+	if err != nil {
+		logger.Noticef(err.Error())
+		return
+	}
+
+	// iterate over all hotplug interfaces
+	for _, iface := range hotplugIfaces {
+		hotplugHandler := iface.(hotplug.Definer)
+
+		// determine device key for the interface; note that interface might provide own device keys.
+		key, err := deviceKey(defaultKey, devinfo, iface)
+		if err != nil {
+			logger.Noticef(err.Error())
+			continue
+		}
+
+		spec := hotplug.NewSpecification()
+		if hotplugHandler.HotplugDeviceDetected(devinfo, spec) != nil {
+			logger.Noticef("Failed to process hotplug event by the rule of interface %q: %s", iface.Name(), err)
+			continue
+		}
+		slotSpecs := spec.Slots()
+		if len(slotSpecs) == 0 {
+			continue
+		}
+
+		if !validateDeviceKey(key) {
+			logger.Debugf("No valid device key provided by interface %s, device %q ignored", iface.Name(), devinfo.DeviceName())
+			continue
+		}
+
+		if !hotplugFeature {
+			logger.Noticef("Hotplug 'add' event for device %q (interface %q) ignored, enable experimental.hotplug", devinfo.DevicePath(), iface.Name())
+			continue
+		}
+
+		logger.Debugf("HotplugDeviceAdded: %s (default key: %q, devname %s, subsystem: %s)", devinfo.DevicePath(), defaultKey, devinfo.DeviceName(), devinfo.Subsystem())
```

Comment:
Since it's per interface, maybe adding some iface name would be useful (just in case they have the same keys or end up using default).